### ISSUE TITLE
[Bug][ROCm] Sparse indexer clears the full max_model_len logits buffer every decode step: 17% throughput on a 1M context

### ISSUE THREAD

**Issue by stefanskiasan** [vllm-project/vllm#55373]
### Your current environment

ROCm 7.2.3, MI355X (gfx950), TP4, vLLM main.
GLM-5.3 (`glm_moe_dsa`, hidden 6144, 256 experts), `--max-model-len 1048576`,
MTP k=3, 64 concurrent requests.

### 🐛 Describe the bug

On ROCm, the DeepSeek-V3.2 style sparse indexer clears its entire logits buffer
on every decode step, even though the consumer only ever reads up to the real
sequence lengths. The buffer is sized `(B * next_n, max_model_len)`, so with a
1M context window this is a large amount of pure overhead.

`vllm/v1/attention/ops/rocm_aiter_mla_sparse.py`, both ROCm paths:

```python
# HIP indexer kernel path
out_logits.fill_(float("-inf"))
...
# deepgemm path
out_logits.nan_to_num_(float("-inf"))
```

Profiled with the torch profiler, 20 decode iterations:

| load | kernel | share of GPU time |
|---|---|---|
| 1 request | `__amd_rocclr_fillBufferAligned` | **4.5 %** |
| 64 concurrent | `nan_to_num` (vectorized_elementwise) | **6.4 %** |

Sizing: at 64 requests with MTP k=3 the buffer is `256 x 1048576 x 4 B = 1.07 GB`,
while an 8000-token context uses 8.2 MB of it — **0.76 %**. The arithmetic matches
the measurement: 1.07 GB written at ~4 TB/s is 268 µs, and the profiler shows
239.6 µs per call.

### Why it looks unnecessary

The same caller (`sparse_attn_indexer_kpool.py`) passes **`clean_logits=False`**
on the CUDA path and does not clear at all. That works because the consumer,
`torch.ops._C.top_k_per_row_decode(logits, next_n, seq_lens, ...)`, is given the
real per-request sequence lengths and never reads past them. The region beyond
`seq_lens` is dead on every platform.

ROCm has no equivalent to `clean_logits` and clears unconditionally.

### Measurement with the clearing removed

Same build, same flags, one env switch. Correctness first — a faster run with
worse answers would not be a win:

Two full runs, with the order of the arms reversed in the second so an ordering
artifact would show up:

| | clearing on | clearing off |
|---|---|---|
| GSM8K (300 questions), run 1 | 97.7 % | 98.3 % |
| GSM8K (300 questions), run 2 | 97.3 % | 98.7 % |
| needle-in-haystack (66k ctx) | found | found |
| single stream (median of 6) | 178.5 tok/s | 179.0 |
| throughput, 64 concurrent, run 1 | 2064 / 2256 tok/s | 2425 / 2663 |
| throughput, 64 concurrent, run 2 | 2024 / 2360 tok/s | 2391 / 2687 |
| **mean throughput** | **2176 tok/s** | **2542 tok/s** |

**+16.8 % throughput with accuracy unchanged** (the 1-point GSM8K difference is
noise in both directions). Single-stream is unaffected, which fits: there the
buffer is only 16 MB and clearing it costs little.

### Suggestion

Give the ROCm path the same treatment as CUDA. Either:

1. **Skip the clearing**, matching `clean_logits=False` — simplest, and what the
   numbers above measure; or
2. **Bound it to the batch maximum** if some consumer does rely on the padding.
   Note that `seq_lens.max().item()` would synchronise and is not usable during
   CUDA graph capture, so this needs a host-side bound.

I would rather not guess which of the two is right for every consumer of this
buffer, so I am reporting it as an issue rather than sending a patch straight
away. Happy to prepare either once someone who knows the indexer's contract
weighs in — in particular whether the HIP kernel is guaranteed to write every
position inside the causal window, which is what option 1 relies on.

**Comment by stefanskiasan** [comment]
Follow-up: I said in the issue that I did not know whether the HIP kernel writes
every position inside the causal window, and that this was the reason to report
rather than patch. That turned out to be measurable, so here are the numbers.

Pre-filling the buffer with a sentinel value instead of `-inf`, then counting
survivors inside each row's causal bound:

| ctx | sentinels inside bound | beyond bound | rows differing (control) |
|---|---|---|---|
| 512 | **0** | 24 | 4 (of 4) |
| 2000 | **0** | 24 | 16 (of 16) |
| 8192 | **0** | 24 | 16 (of 16) |
| 30000 | **0** | 24 | 16 (of 16) |

- **No sentinel survives inside the bound**, so the kernel leaves no gaps there.
- The 24 beyond it are exactly `sum(next_n-1..0) * B` — the positions past each
  row's causal limit, which `top_k_per_row_decode` never reads because it is
  given `next_n` and `seq_lens`.
- The last column is the control: two `-inf` runs compared against each other.
  They differ by exactly as much as a sentinel run differs from a `-inf` run, so
  the kernel is simply not bitwise reproducible and the pre-fill value has no
  effect on the result.

One methodological note in case someone repeats this: my first version compared
`[0, ctx)` for every row and reported "24 gaps", which looks alarming until the
per-row causal bound is accounted for. The window is `ctx - next_n + n + 1` for
row `n`, not `ctx`.

So the clearing is safe to skip. Branch with the patch (adds `clean_logits` to
the ROCm path and passes `False` from the same call site as CUDA):
https://github.com/stefanskiasan/vllm/tree/rocm-skip-indexer-logit-clean — I hit
GitHub's PR creation rate limit today and will open it shortly.
